controllers/upgrade_handlers.go

```diff
@@ -41,20 +42,28 @@ var (
 	defaultRebootTimeout = 60 * time.Minute
 )
 
-var isPrePivot = func(r *ImageBasedUpgradeReconciler, ibu *lcav1alpha1.ImageBasedUpgrade) (bool, error) {
-	currentStaterootName, err := r.RPMOstreeClient.GetCurrentStaterootName()
-	if err != nil {
-		return false, err
-	}
-	r.Log.Info("stateroots", "current stateroot:", currentStaterootName, "desired stateroot", getDesiredStaterootName(ibu))
-	return currentStaterootName != getDesiredStaterootName(ibu), nil
+// UpgradeHandler note reconcile.Reconciler interface was used instead of ibu reconciler to avoid circular dependency
+type UpgradeHandler interface {
+	HandleBackup(r reconcile.Reconciler, ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error)
+	HandleRestore(r reconcile.Reconciler, ctx context.Context) (ctrl.Result, error)
+	RebootToNewStateRoot(r reconcile.Reconciler, ibu *lcav1alpha1.ImageBasedUpgrade) error
+	ExportIBUCR(ibu *lcav1alpha1.ImageBasedUpgrade, filePath string) error
+	PostPivot(r reconcile.Reconciler, ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error)
+	PrePivot(r reconcile.Reconciler, ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error)
+	IsPrePivot(r reconcile.Reconciler, ibu *lcav1alpha1.ImageBasedUpgrade) (bool, error)
+}
+
+type UpgHandler struct{}
```

Comment:
Can you add what UpgHandler needs from reconciler (BackupRestore, log, etc) to this instead?

Reply:
Yeah i thought of that! But there are things that we need directly from reconciler but are shared by other handlers...e.g executor. So just taking out (e.g) BackRestore and leaving out few others didn't really make sense as this will also open us to ever increasing number of params imo...

Reply:
I meant everything it needs. Can't we just initialize it with everything it needs in `main.go` like everything else, you can pass in executor there, too.
```
	r := rr.(*ImageBasedUpgradeReconciler)
	ctrlResult, err := r.UpgradeHandler.HandleBackup(r, ctx, ibu)
```
because this does not look good to me.

Reply:
Like we spoke about....will get rid mockgen in upgradehandler and refactor from the top level . That should get rid of the ugly part but will lose some of the mockgen functionality. Updating the code now